### ISSUE TITLE
Add possibility to checkout pull request locally 

### ISSUE THREAD

**Issue by hhenkel** [gogits/gogs#1655]
There are various reason why one might want to checkout a pull request locally.
- When there is conflicting pull request it should be possible to checkout the request itself to resolve the issue locally.
- It's enabling people to configure there CI solution to checkout and test the pull request (and comment the result back to gogs once the api is ready).
- The original author might have deleted his fork of the repo, but one might still want to accept the pull request.

This is also possible on other solutions like github - [merging a pull request](https://help.github.com/articles/merging-a-pull-request/#checking-out-a-pull-request-locally) / [checking out pull requests locally](https://help.github.com/articles/checking-out-pull-requests-locally/) or [gitblit](http://gitblit.com/) which has something called [tickets](http://gitblit.com/tickets_overview.html).

**Comment by Unknwon** [comment]
Thanks your feedback!

**Comment by hhenkel** [comment]
Just recognized that it might be worth mentioning, that it is possible to show all the remote refs of a repository via "git ls-remote". In the case of github (gogs repo) this shows refs to all pulls (issues?) besides the obvious stuff:

```
$ git ls-remote
From https://github.com/gogits/gogs.git
eec06fb3df7a9dc30f758bb1a100b5f1e8e01ab9        HEAD
bd5dc626e82e18d3e619d918e579dc130edcd1fa        refs/heads/develop
10490f1d372ee538231ce3a378e3dd0ec23fef7b        refs/heads/feature/comment
f2ab68e2201e7ddc83fc6363caadb4893bd42fd7        refs/heads/git
eec06fb3df7a9dc30f758bb1a100b5f1e8e01ab9        refs/heads/master
22bb5104c8a0d585ea0e538410af1044bc2c1505        refs/pull/100/head
e63624ca0ecf1052f25e11ff06c6e6b52c0deafb        refs/pull/1001/head
04565ed5deb502c70a964365b1531d681901ae35        refs/pull/1001/merge
698ce99a646f0b9f4224abea5fbdb34278cc305a        refs/pull/1005/head
d98516899e4344a11fa6230d3eacbe8fb5686b30        refs/pull/1007/head
6f094b9829b485fdb88e439263b9bb8bcd2429e5        refs/pull/1007/merge
97b57f085d7231dab99a283600e6851ed14fe30e        refs/pull/1019/head
767e50b2648103b39430219e596ce6a82f0ddfd1        refs/pull/1019/merge
...
...
...
58436b5ea51cb5e09fb244331d7a9acf70258063        refs/tags/v0.7.0
d37da1f3929f94703d7c7f19f60288109f95756d        refs/tags/v0.7.19
253513cedda72939ec4d496a3325c654659cc15f        refs/tags/v0.7.22
b5f6206a659eaa76382774675f2e3f2664d0f501        refs/tags/v0.7.33
d1e28ac013c2f562939025e6a031671a5a70dd52        refs/tags/v0.7.6
```

With this is should be possible to checkout a pull request.

**Comment by hhenkel** [comment]
@Unknwon I spent a little time searching on the net and I found this:
"refs/pull/2/head is a reference to the head commit of your pull request, whilst refs/pull/2/merge is a reference to the result of the automatic merge that GitHub does." [Source](https://www.finalbuilder.com/resources/blogs/postid/700/building-github-pull-requests-with-continua-ci)

**Comment by Unknwon** [comment]
@hhenkel Thanks!

**Comment by Unknwon** [comment]
@hhenkel first working version is now on `develop` branch, please do all kinds of crazy things to help test!

**Comment by hhenkel** [comment]
Hmm, not sure what I'm doing wrong or if a expect to much, maybe @mhartkorn can enlighten me?

I got a new repo which was initialized with the README. I forked the repo did some minor change to the README file and opened up a pull request against the initial repo.

I'm not able to see the pull via `git ls-remote`:
```
$ git ls-remote
From https://git.linux.example.com/gogs/PuppetModules/test.git
c413d30292f192506c5905d990cabcca94fd119f        HEAD
c413d30292f192506c5905d990cabcca94fd119f        refs/heads/master
```

According to the github documentation it should be possible to checkout a pull request locally by doing:

```
git fetch origin pull/ID/head:BRANCHNAME
git checkout BRANCHNAME
```

Which works for the pull requests on github:
```
$ git fetch origin pull/2676/head
remote: Counting objects: 4, done.
remote: Compressing objects: 100% (4/4), done.
remote: Total 4 (delta 0), reused 0 (delta 0), pack-reused 0
Unpacking objects: 100% (4/4), done.
From https://github.com/gogits/gogs
 * branch            refs/pull/2676/head -> FETCH_HEAD

$ git checkout FETCH_HEAD
Note: checking out 'FETCH_HEAD'.

You are in 'detached HEAD' state. You can look around, make experimental
changes and commit them, and you can discard any commits you make in this
state without impacting any branches by performing another checkout.

If you want to create a new branch to retain commits you create, you may
do so (now or later) by using -b with the checkout command again. Example:

  git checkout -b new_branch_name

HEAD is now at 331732b... tell macaron to use auth head
```

For gogs I get the following:
```
$ git fetch origin pull/1/head
fatal: Couldn't find remote ref pull/1/head
Unexpected end of command stream
```

**Comment by mhartkorn** [comment]
Is the project older? It will not work for old Pull Requests that were created before Gogs 0.8.40.

**Comment by hhenkel** [comment]
No, I tested with version 0.8.41.0221 (builded it yesterday morning) of gogs and a fresh repository.

**Comment by mhartkorn** [comment]
Bug confirmed here on 0.8.43.0223: https://try.gogs.io/mattddowney/go-get-test/pulls/1 I cannot look into it until tomorrow.

**Comment by Unknwon** [comment]
I think the problem is current code only create/update and push commits to base repo when user push new code, not when create/close the PR.

**Comment by Unknwon** [comment]
Look at code of how testing patch works, may help.

**Comment by mhartkorn** [comment]
@Unknwon I tried to patch i, should I reopen a new PR?

@hhenkel In the meantime you can try out this patch: https://github.com/mhartkorn/gogs/commit/51f15880d1f25f9d40d93762c5dcb52c2dccac14.patch

**Comment by Unknwon** [comment]
@mhartkorn Yes, please try a new PR, so I can directly help test.

**Comment by hhenkel** [comment]
@mhartkorn I just rebuilded the current develop branch (based on d324500959c06e975921790f8770aa5d1bdf2344) with you patch included. I still can't see that the behavior has changed...?

**Comment by hhenkel** [comment]
Ah, my bad...let me retry.

**Comment by Unknwon** [comment]
@hhenkel can you test again with `develop` branch?

**Comment by hhenkel** [comment]
@Unknwon yes, sorry for the delay, been busy the last few days. I just gave it a try and it seems there is currently an error in develop:

```
$ go build -tags "sqlite pam cert"
# github.com/gogits/gogs/models
models/repo.go:376: undefined: git.PullRemoteOptions
```

**Comment by mhartkorn** [comment]
@hhenkel The patch is also in `master`.

**Comment by Unknwon** [comment]
@hhenkel you need to update dependencies... `go get -u github.com/gogits/git-module`

**Comment by hhenkel** [comment]
@Unknwon / @mhartkorn sorry, was to busy and distracted at work.

I just gave it a try with the current develop branch and it looks good to me so far:

```
$ git ls-remote
From http://localhost:3000/test/upstream.git
78501add6739796f12d728de95cb763b08d28d96        HEAD
78501add6739796f12d728de95cb763b08d28d96        refs/heads/master
76112873d46518bae7b74c05c9b69ad6209148c3        refs/pull/1/head
```

```
$ git fetch origin pull/1/head
remote: Counting objects: 3, done.
remote: Total 3 (delta 0), reused 0 (delta 0)
Unpacking objects: 100% (3/3), done.
From http://localhost:3000/test/upstream
 * branch            refs/pull/1/head -> FETCH_HEAD
```

**Comment by boeboe** [comment]
Any update on this PR?

**Comment by Unknwon** [comment]
@boeboe what PR?

**Comment by boeboe** [comment]
Never mind... I meant weather there was a plan to expose PR meta data trough refs, a needed feature for pipeline solutions that want builds triggered on PR and checkout the PR for testing.

**Comment by Unknwon** [comment]
@boeboe What kind of "meta data" you're looking for?

**Comment by boeboe** [comment]
The comment and commit messages linked with a PR.

**Comment by Unknwon** [comment]
Oh... for that, I think you need to watch out this PR: https://github.com/gogits/gogs/issues/2253...

**Comment by boeboe** [comment]
Thanks @Unknwon
